Import face region metadata

kvalev opened the pull request:
Read the most popular face region formats from exif/xmp and import them into PhotoPrism. The following formats are supported: IPTC, MWG, WLPG

During indexing, the file metadata information will be read and if it contains exif face regions, they will be imported into photoprism. No face clustering or auto-tagging will be performed, rather the information from the exif metadata will be imported as-is.

During indexing, face regions will be de-duplicated (in case the same face is described in multiple formats).
Both "normal" as well as rotated images are supported (all 8 different orientations).
In case a face region is ignored during import, the reason will be included in the logs (e.g. malformed data or unsupported input).

The following scenarios are supported and have been tested:

- [x] index the library from scratch using this pr: all exif face regions are successfully imported
- [x] reindex the library (after 1.): all exif face regions are skipped, as they have been imported during index
- [x] reset face data (after 1.) and reindex the library: all exif face regions are successfully imported
- [x] index the library from scratch using current preview branch and then reindex the library using this pr: all exif face regions are successfully imported
- [x] index the library from scratch using current preview branch, manually name some of the detected faces which also have exif face regions and then reindex the library using this pr: the manually named markers will be kept as-is (even if the manual name does not match the exif name) and the rest of the exif face regions will be imported; this might result in multiple subjects for the same person (if the names does not match), but that's being detected by photoprism `faces: marker [m] face [F] has ambiguous subjects [j1] <> [j2], subject source meta`

Sources:
- https://github.com/exiftool/exiftool/blob/master/config_files/convert_regions.config
- https://github.com/bpatrik/pigallery2/blob/dc7bc032cec8156bb77416896bc271497e89bc8d/src/backend/model/threading/MetadataLoader.ts#L318-L352
- https://github.com/bpatrik/ts-exif-parser/blob/bd6bb335b1923ff55be5dd42db434bde455c5204/lib/ExifData.ts#L3-L12
- https://www.iptc.org/std/photometadata/documentation/userguide/
- https://s3.amazonaws.com/software.tagthatphoto.com/docs/mwg_guidance.pdf
- https://www.daveperrett.com/articles/2012/07/28/exif-orientation-handling-is-a-ghetto/

related to https://github.com/photoprism/photoprism/issues/747, https://github.com/photoprism/photoprism/issues/554, https://github.com/photoprism/photoprism/issues/1570

kvalev committed:
People: Read MWG and WLPG face region metadata from exif

kvalev committed:
People: Read IPTC face region metadata from exif

kvalev committed:
People: Unique exif face regions

The same/similar face region might de declared multiple time, e.g. in different
formats. A check has been added which will ensure that all face regions do not
conflict with one another.

rickysarraf commented:
@kvalev Is this PR supposed to update `People` records based on the facial embeddings written by Digikam in EXIF format ?
So far, I have it built and tested.

I get below debug messages:

```
time="2023-05-05T07:53:40Z" level=debug msg="mp: mp4 was already extracted from motion photo Pictures/2022_Processed/PXL_20221002_140439147.MP.mp4"
time="2023-05-05T07:53:40Z" level=debug msg="index: created motion photo video for PXL_20221002_140439147.MP.mp4"              
time="2023-05-05T07:53:41Z" level=debug msg="index: PXL_20221002_140439147.MP.jpg was modified (new size 2710364, old size 2710364, new timestamp 1683106767, old timestamp 1682010632)"
time="2023-05-05T07:53:41Z" level=debug msg="media: opened Pictures/2022_Processed/PXL_20221002_140439147.MP.jpg [32 MB]"             
time="2023-05-05T07:53:42Z" level=info msg="media: created 8 thumbnails for Pictures/2022_Processed/PXL_20221002_140439147.MP.jpg [1.339991321s]"
time="2023-05-05T07:53:42Z" level=debug msg="faces: found face region metadata [{3264 2448 0 722-1083-408 [] [] []}]"                              
time="2023-05-05T07:53:42Z" level=info msg="index: found 1 face in PXL_20221002_140439147.MP.jpg [172.65307ms]"                                    
time="2023-05-05T07:53:45Z" level=debug msg="search: updated index by photo id [4.135705ms]"                                                                                                      
time="2023-05-05T07:53:45Z" level=debug msg="photo: Pictures/2022_Processed/PXL_20221002_140439147 title based on 1 person ('Anshika Raj AAA Sarraf')"                                  
time="2023-05-05T07:53:46Z" level=debug msg="search: updated index by photo id [4.111936ms]"                                                                                                      
time="2023-05-05T07:53:46Z" level=debug msg="search: updated index by photo id [4.023281ms]"                                                                                                      
```

But I'm not sure if it is updating the `People` records accordingly.

kvalev commented:
@rickysarraf Yeah thats exactly the goal of this PR. Right now I have committed only the EXIF parsing part, so the faces are only printed (https://github.com/kvalev/photoprism/pull/59/files#diff-9d687d9e5abd223347ebb1538e986d34bee6dbe22f7f25075ec6dea1e6bc0a3bR327-R329), but not imported yet.

The rest of the PR is actually that I have implemented locally already, but there are a couple of caveats, which is the reason why this PR stayed open so long:

1. Due to the way photoprism works, it will only consider a region to be a face if the neural net it uses detects it as a face. This means that if you have a region which is not detected as face by the facenet, the region will be ignored and it will not appear in photoprism afterwards. There might be a way to make it work, but it will probably be quite convoluted.

2. Even if we only import the detected face regions, the next problem is the name associated with the EXIF region. Per default, when photoprism detects a face, it will also cluster it and often enough it will find a proper cluster for it and as such it will automatically assign a name. However the automatically assigned name might differ from the EXIF name - either because of a different spelling of the name or maybe because photoprism assigned the face to the wrong person.

Those are I think the biggest question marks around the implementation. I personally dont have any pictures with face regions in EXIF, so for me its a bit difficult to judge what would be the expectation for this feature. If you can provide me with a bit of feedback that would be helpful.

rickysarraf commented:
> @rickysarraf Yeah thats exactly the goal of this PR. Right now I have committed only the EXIF parsing part, so the faces are only printed (https://github.com/kvalev/photoprism/pull/59/files#diff-9d687d9e5abd223347ebb1538e986d34bee6dbe22f7f25075ec6dea1e6bc0a3bR327-R329), but not imported yet.
> 
> The rest of the PR is actually that I have implemented locally already, but there are a couple of caveats, which is the reason why this PR stayed open so long:
> 
> 1. Due to the way photoprism works, it will only consider a region to be a face if the neural net it uses detects it as a face. This means that if you have a region which is not detected as face by the facenet, the region will be ignored and it will not appear in photoprism afterwards. There might be a way to make it work, but it will probably be quite convoluted.


Digikam's facial recognition is more robust than any of the other tools I've used. So I'd rather brute force a face, for embedding read from Digikam.

Also, note that you have a manual face tagging feature implemented in your fork. And in my tests so far, all manually tagged (low quality faces too) faces were detected well by PhotoPrism.

Given PhotoPrism terrible performance with Facial Recognition, I am exploring possible integrations with 3rd party tools. And this PR came as the highest hope.

Hope is to:

* Reliably detect/recognize faces in Digikam, which is also very very flexible in other regards to tagging.
* Write metadata to Photos directly in their EXIF region.
* Re-scan PhotoPrism library (with your patches) and be able to read all facial embeddings
* Make PhotoPrism populate its fields from those EXIF information.

> 2. Even if we only import the detected face regions, the next problem is the name associated with the EXIF region. Per default, when photoprism detects a face, it will also cluster it and often enough it will find a proper cluster for it and as such it will automatically assign a name. However the automatically assigned name might differ from the EXIF name - either because of a different spelling of the name or maybe because photoprism assigned the face to the wrong person.
> 

Well. One has to make a core assumption somewhere. For me, it has been:

* Digikam Names == PhotoPrism Names

> Those are I think the biggest question marks around the implementation. I personally dont have any pictures with face regions in EXIF, so for me its a bit difficult to judge what would be the expectation for this feature. If you can provide me with a bit of feedback that would be helpful.

Please feel free to always ping me for any such feedback. It has been a pleasure working with your fork. It has also helped me slightly improve my understanding of this Golang based project.

kvalev commented:
Would you be willing to test this feature? I dont have any tagged faces in EXIF, so I can only speculate what problems might occur. It will be helpful to have someone with a lot of EXIF faces test it, but there is a risk of messing up your library. Might make sense to have a separate instance only for testing it, but that would require a bit more effort on your side..

rickysarraf commented:
> Would you be willing to test this feature? I dont have any tagged faces in EXIF, so I can only speculate what problems might occur. It will be helpful to have someone with a lot of EXIF faces test it, but there is a risk of messing up your library. Might make sense to have a separate instance only for testing it, but that would require a bit more effort on your side..

Yes. Please. I can help with the testing. I already have 2 instances set up here locally

kvalev committed:
People: Improve logging then parsing exif face regions

kvalev committed:
People: Create faces, subjects and markers from exif face regions

During indexing, the metadata information will be read and if it contains exif
face regions, they will be imported into photoprism.
No face clustering or auto-tagging will be performed, rather the information
from the exif metadata will be imported as-is.

kvalev commented:
@rickysarraf and anyone else interested - this feature is now in a decent shape and can be tested. However please create a separate Photoprism instance, as I have only few files with exif face data, so my testing so far has been somewhat limited. As such, there are probably some cases and scenarios that I missed and might lead to a messed up Photoprism face data.

rickysarraf commented:
> @rickysarraf and anyone else interested - this feature is now in a decent shape and can be tested. However please create a separate Photoprism instance, as I have only few files with exif face data, so my testing so far has been somewhat limited. As such, there are probably some cases and scenarios that I missed and might lead to a messed up Photoprism face data.

Thank you. I'm going to test it later today and provide you with my test impressions.

rickysarraf commented:
> > @rickysarraf and anyone else interested - this feature is now in a decent shape and can be tested. However please create a separate Photoprism instance, as I have only few files with exif face data, so my testing so far has been somewhat limited. As such, there are probably some cases and scenarios that I missed and might lead to a messed up Photoprism face data.
> 
> Thank you. I'm going to test it later today and provide you with my test impressions.

@kvalev Please find the 1st round of logs. These were run on a limited set of photos, in the interest of time. I will follow-up with more logs from a larger set. Meanwhile, if you think more verbosity is needed, then please mention and I'll try to enable it.

https://people.debian.org/~rrs/tmp/face-embeddings-folder-Priya.log.gz

rickysarraf commented:
> > > @rickysarraf and anyone else interested - this feature is now in a decent shape and can be tested. However please create a separate Photoprism instance, as I have only few files with exif face data, so my testing so far has been somewhat limited. As such, there are probably some cases and scenarios that I missed and might lead to a messed up Photoprism face data.
> > 
> > 
> > Thank you. I'm going to test it later today and provide you with my test impressions.
> 
> @kvalev Please find the 1st round of logs. These were run on a limited set of photos, in the interest of time. I will follow-up with more logs from a larger set. Meanwhile, if you think more verbosity is needed, then please mention and I'll try to enable it.
> 
> https://people.debian.org/~rrs/tmp/face-embeddings-folder-Priya.log.gz

One cool new feature, which I had hoped, and which this PR implements, is that I had a couple of images with multiple face embeddings, and names assigned to them in Digikam. Written directly into the picture files, in EXIF data.

I was quite happy to see 2 new `People` listed in `Recognized` tab under `People`. The names assigned to the faces in Digikam, were nicely populated in PhotoPrism. Note that these names were never assigned in PhotoPrism. 

So, with this PR:
* PhotoPrism read those embeddings
* Created respective person in PhotoPrism's database
* Assigned the face to the person in PhotoPrism.
* It also took Named Face Embeddings, which already had a named entry (same name in Digikam and PhotoPrism) in PhotoPrism, and updated its record appropriately.

rickysarraf commented:
@kvalev Here's another log, run on a larger set of pictures, including lots of collages. So there are more ~~faces~~ facial embeddings in this log, compared to the previous one.

https://people.debian.org/~rrs/tmp/zMisc-set.logs.gz

* People count nicely increased from 6 to 37

kvalev commented:
@rickysarraf Thanks for testing the feature! Happy that a lot of data is now being correctly imported, but I'm also seeing a lot of errors in the log. The most prominent root cause seems one seems to be: 
https://github.com/kvalev/photoprism/blob/4b96a502774c71071b6a945e6d83113ba5090e94/internal/photoprism/index_mediafile.go#L364

Do you mind sending me few of the filenames that appear in the log, so I can debug it locally? Especially some of those images that have to be upscaled first? (`IMG-20170712-WA0008_1499861497662~2.jpg` might be quite interesting as well, but I dont know how sensitive the photo is.)

One more question: Did you test it in a fresh instance without any tagged faces (either manually or detected by photoprism)?

rickysarraf commented:
> Do you mind sending me few of the filenames that appear in the log, so I can debug it locally? Especially some of those images that have to be upscaled first? (`IMG-20170712-WA0008_1499861497662~2.jpg` might be quite interesting as well, but I dont know how sensitive the photo is.)
>

@kvalev I have privately sent you an email which has a link to a bundle of photos, based on the criteria you pointed in the logs.
 
> One more question: Did you test it in a fresh instance without any tagged faces (either manually or detected by photoprism)?

It isn't a fresh instance entirely but majority of the data was reset. I had only populated 4 People faces in my testing instance, for the primary family members. Every new face (new count at 42 right now) was read from EXIF embeddings written by Digikam, and then populated into PhotoPrism via this PR.

kvalev committed:
People: Improve logging for exif face regions

kvalev committed:
People: Skip already indexed face regions

kvalev commented:
@rickysarraf Thanks! 

I investigated it in a local fresh instance and it seems the reason for the many error logs was one small missing check. I have updated the PR, so if you can test it when you have the time that would be of great help. There should be almost no errors in the logs from now on (independent of whether you do a fresh index or a re-index).

However based on the logs you sent me I identified few pictures where Photoprism's face recognition chokes up, which is a prerequisite even for EXIF face regions, and as such those faces will not be imported. I am not sure how to handle those to be honest, but if you want, you can send me 1 or 2 sample photos to play around with - you can identify them by the log statement `unexpected embeddings count`.

rickysarraf commented:
> @rickysarraf Thanks!
> 
> I investigated it in a local fresh instance and it seems the reason for the many error logs was one small missing check. I have updated the PR, so if you can test it when you have the time that would be of great help. There should be almost no errors in the logs from now on (independent of whether you do a fresh index or a re-index).
> 
> However based on the logs you sent me I identified few pictures where Photoprism's face recognition chokes up, which is a prerequisite even for EXIF face regions, and as such those faces will not be imported. I am not sure how to handle those to be honest, but if you want, you can send me 1 or 2 sample photos to play around with - you can identify them by the log statement `unexpected embeddings count`.

Thank you for the new changes @kvalev 

Please find below the link to logs from this test.
https://people.debian.org/~rrs/tmp/2020-PP.logs.gz

I have emailed you separately for pics that were reported problematic for facial embeddings.

kvalev committed:
People: Additional MWG face region validation and logging

As only normalized NWG face regions are supported, all other region types should
be refected and it should be logged why the region was discarded.

kvalev committed:
People: Handle face regions in rotated images

If an image is rotated (meaning it does not have a "normal" orientation), all of
the face regions' information (such as x, y position and width, heigh) are
rotated as well.
PhotoPrism however expects that all face markers are in "normal", non-rotated
coordinates, so we need to convert the face regions if the image is rotated.

kvalev committed:
People: Improve debug logs

kvalev commented:
@rickysarraf I figured out the problem - it turned out that rotated images (and face regions) require a bit of special handling. I have fixed that and now all of the sample photos you sent me are imported correctly (with the exception of one very small face in a photo frame, but I think that would be acceptable). Can you give it another go, with even more data if possible, and let me know if there are any error logs.

kvalev committed:
People: Update face region parsing and normalization documentation

rickysarraf commented:
> @rickysarraf I figured out the problem - it turned out that rotated images (and face regions) require a bit of special handling. I have fixed that and now all of the sample photos you sent me are imported correctly (with the exception of one very small face in a photo frame, but I think that would be acceptable). Can you give it another go, with even more data if possible, and let me know if there are any error logs.

I have run a wider import/scan. So far, below quick error has caught my attention.

```
photoprism_1  | 2023-05-23T14:10:47.839273939Z time="2023-05-23T14:10:47Z" level=debug msg="import: created 62f27aab190bee8344afcceed3c2c40e203356c7_exiftool.json"                                                              [0/89799]
photoprism_1  | 2023-05-23T14:10:47.869143409Z time="2023-05-23T14:10:47Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_193909.jpg"
photoprism_1  | 2023-05-23T14:10:47.872872087Z time="2023-05-23T14:10:47Z" level=info msg="media: 20191211_193909.jpg was taken at 2019-12-11 14:09:09 +0000 UTC (meta)"
photoprism_1  | 2023-05-23T14:10:47.915403431Z time="2023-05-23T14:10:47Z" level=info msg="import: moving main jpg file Digikam_TO_BE_FILTERED/20191211_193909.jpg to 2019/12/20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:47.915492376Z time="2023-05-23T14:10:47Z" level=debug msg="failed renaming file, fallback to copy and delete: rename /photoprism/import/Digikam_TO_BE_FILTERED/20191211_193909.jpg /photoprism/originals/2019/12/20191211_140909_EF5BAC99.jpg: invalid cross-device link"
photoprism_1  | 2023-05-23T14:10:47.999226650Z time="2023-05-23T14:10:47Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:48.680906540Z time="2023-05-23T14:10:48Z" level=debug msg="media: opened 2019/12/20191211_140909_EF5BAC99.jpg [32 MB]"
photoprism_1  | 2023-05-23T14:10:49.254782000Z time="2023-05-23T14:10:49Z" level=info msg="media: created 8 thumbnails for 2019/12/20191211_140909_EF5BAC99.jpg [1.331534203s]"
photoprism_1  | 2023-05-23T14:10:49.265964557Z time="2023-05-23T14:10:49Z" level=debug msg="media: created no new thumbnails for 2019/12/20191211_140909_EF5BAC99.jpg [369.057µs]"
photoprism_1  | 2023-05-23T14:10:49.297630506Z time="2023-05-23T14:10:49Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:49.301049469Z time="2023-05-23T14:10:49Z" level=debug msg="index: found face region metadata in 2019/12/20191211_140909_EF5BAC99.jpg ([anshika-raj-sarraf-1967-1182-401 priya-raj-ppp-sarraf-1527-525-414])"
photoprism_1  | 2023-05-23T14:10:49.301128151Z time="2023-05-23T14:10:49Z" level=debug msg="index: processing face region anshika-raj-sarraf-1967-1182-401"
photoprism_1  | 2023-05-23T14:10:49.301434042Z time="2023-05-23T14:10:49Z" level=debug msg="index: calculated face score 9"
photoprism_1  | 2023-05-23T14:10:50.419900267Z time="2023-05-23T14:10:50Z" level=debug msg="markers: added face mrv47e225115wz1o for file ''"
photoprism_1  | 2023-05-23T14:10:50.419993723Z time="2023-05-23T14:10:50Z" level=error msg="markers: did not save new marker for 2019/12/20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:50.420393417Z time="2023-05-23T14:10:50Z" level=debug msg="index: processing face region priya-raj-ppp-sarraf-1527-525-414"
photoprism_1  | 2023-05-23T14:10:50.420439082Z time="2023-05-23T14:10:50Z" level=debug msg="index: calculated face score 9"
photoprism_1  | 2023-05-23T14:10:51.482661598Z time="2023-05-23T14:10:51Z" level=debug msg="markers: added face mrv47e319j8r7yxl for file ''"
photoprism_1  | 2023-05-23T14:10:51.482749385Z time="2023-05-23T14:10:51Z" level=error msg="markers: did not save new marker for 2019/12/20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:51.607376332Z time="2023-05-23T14:10:51Z" level=info msg="index: found 2 faces in 20191211_140909_EF5BAC99.jpg [124.007601ms]"
photoprism_1  | 2023-05-23T14:10:53.451181853Z time="2023-05-23T14:10:53Z" level=debug msg="photo: changing location of 2019/12/20191211_140909_EF5BAC99 from zz to s2:3bae0b219dfc"
photoprism_1  | 2023-05-23T14:10:53.778594631Z time="2023-05-23T14:10:53Z" level=debug msg="photo: 2019/12/20191211_140909_EF5BAC99 title based on label 'The Empyrean'"
photoprism_1  | 2023-05-23T14:10:53.778671318Z time="2023-05-23T14:10:53Z" level=debug msg="photo: 2019/12/20191211_140909_EF5BAC99 has new title 'The Empyrean / Jagadenahalli / 2019' [2.027553ms]"
photoprism_1  | 2023-05-23T14:10:56.264843830Z time="2023-05-23T14:10:56Z" level=debug msg="search: updated index by photo id [360.884629ms]"
photoprism_1  | 2023-05-23T14:10:56.270272165Z time="2023-05-23T14:10:56Z" level=debug msg="index: updated yaml file 20191211_140909_EF5BAC99.yml"
photoprism_1  | 2023-05-23T14:10:56.270360248Z time="2023-05-23T14:10:56Z" level=info msg="import: added main jpg file 2019/12/20191211_140909_EF5BAC99.jpg"
photoprism_1  | 2023-05-23T14:10:56.323644281Z time="2023-05-23T14:10:56Z" level=debug msg="exiftool: extracting metadata from Digikam_TO_BE_FILTERED/20191211_193912.jpg"
photoprism_1  | 2023-05-23T14:10:56.928100969Z time="2023-05-23T14:10:56Z" level=debug msg="import: created 75076f4af309e03970a8dc93a645b6d59bf10139_exiftool.json"
photoprism_1  | 2023-05-23T14:10:56.957007358Z time="2023-05-23T14:10:56Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_193912.jpg"
photoprism_1  | 2023-05-23T14:10:56.960946242Z time="2023-05-23T14:10:56Z" level=info msg="media: 20191211_193912.jpg was taken at 2019-12-11 14:09:12 +0000 UTC (meta)"
photoprism_1  | 2023-05-23T14:10:56.964104854Z time="2023-05-23T14:10:56Z" level=info msg="import: moving main jpg file Digikam_TO_BE_FILTERED/20191211_193912.jpg to 2019/12/20191211_140912_036DA417.jpg"

```

I'm quite frequently getting this error: `did not save new marker for`

rickysarraf commented:
> > @rickysarraf Thanks!
> > I investigated it in a local fresh instance and it seems the reason for the many error logs was one small missing check. I have updated the PR, so if you can test it when you have the time that would be of great help. There should be almost no errors in the logs from now on (independent of whether you do a fresh index or a re-index).
> > However based on the logs you sent me I identified few pictures where Photoprism's face recognition chokes up, which is a prerequisite even for EXIF face regions, and as such those faces will not be imported. I am not sure how to handle those to be honest, but if you want, you can send me 1 or 2 sample photos to play around with - you can identify them by the log statement `unexpected embeddings count`.
> 
> Thank you for the new changes @kvalev
> 
> Please find below the link to logs from this test. https://people.debian.org/~rrs/tmp/2020-PP.logs.gz



> > @rickysarraf I figured out the problem - it turned out that rotated images (and face regions) require a bit of special handling. I have fixed that and now all of the sample photos you sent me are imported correctly (with the exception of one very small face in a photo frame, but I think that would be acceptable). Can you give it another go, with even more data if possible, and let me know if there are any error logs.
> 
> I have run a wider import/scan. So far, below quick error has caught my attention.
> 
> ```
> photoprism_1  | 2023-05-23T14:10:47.839273939Z time="2023-05-23T14:10:47Z" level=debug msg="import: created 62f27aab190bee8344afcceed3c2c40e203356c7_exiftool.json"                                                              [0/89799]
> photoprism_1  | 2023-05-23T14:10:47.869143409Z time="2023-05-23T14:10:47Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_193909.jpg"
> photoprism_1  | 2023-05-23T14:10:47.872872087Z time="2023-05-23T14:10:47Z" level=info msg="media: 20191211_193909.jpg was taken at 2019-12-11 14:09:09 +0000 UTC (meta)"
> photoprism_1  | 2023-05-23T14:10:47.915403431Z time="2023-05-23T14:10:47Z" level=info msg="import: moving main jpg file Digikam_TO_BE_FILTERED/20191211_193909.jpg to 2019/12/20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:47.915492376Z time="2023-05-23T14:10:47Z" level=debug msg="failed renaming file, fallback to copy and delete: rename /photoprism/import/Digikam_TO_BE_FILTERED/20191211_193909.jpg /photoprism/originals/2019/12/20191211_140909_EF5BAC99.jpg: invalid cross-device link"
> photoprism_1  | 2023-05-23T14:10:47.999226650Z time="2023-05-23T14:10:47Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:48.680906540Z time="2023-05-23T14:10:48Z" level=debug msg="media: opened 2019/12/20191211_140909_EF5BAC99.jpg [32 MB]"
> photoprism_1  | 2023-05-23T14:10:49.254782000Z time="2023-05-23T14:10:49Z" level=info msg="media: created 8 thumbnails for 2019/12/20191211_140909_EF5BAC99.jpg [1.331534203s]"
> photoprism_1  | 2023-05-23T14:10:49.265964557Z time="2023-05-23T14:10:49Z" level=debug msg="media: created no new thumbnails for 2019/12/20191211_140909_EF5BAC99.jpg [369.057µs]"
> photoprism_1  | 2023-05-23T14:10:49.297630506Z time="2023-05-23T14:10:49Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:49.301049469Z time="2023-05-23T14:10:49Z" level=debug msg="index: found face region metadata in 2019/12/20191211_140909_EF5BAC99.jpg ([anshika-raj-sarraf-1967-1182-401 priya-raj-ppp-sarraf-1527-525-414])"
> photoprism_1  | 2023-05-23T14:10:49.301128151Z time="2023-05-23T14:10:49Z" level=debug msg="index: processing face region anshika-raj-sarraf-1967-1182-401"
> photoprism_1  | 2023-05-23T14:10:49.301434042Z time="2023-05-23T14:10:49Z" level=debug msg="index: calculated face score 9"
> photoprism_1  | 2023-05-23T14:10:50.419900267Z time="2023-05-23T14:10:50Z" level=debug msg="markers: added face mrv47e225115wz1o for file ''"
> photoprism_1  | 2023-05-23T14:10:50.419993723Z time="2023-05-23T14:10:50Z" level=error msg="markers: did not save new marker for 2019/12/20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:50.420393417Z time="2023-05-23T14:10:50Z" level=debug msg="index: processing face region priya-raj-ppp-sarraf-1527-525-414"
> photoprism_1  | 2023-05-23T14:10:50.420439082Z time="2023-05-23T14:10:50Z" level=debug msg="index: calculated face score 9"
> photoprism_1  | 2023-05-23T14:10:51.482661598Z time="2023-05-23T14:10:51Z" level=debug msg="markers: added face mrv47e319j8r7yxl for file ''"
> photoprism_1  | 2023-05-23T14:10:51.482749385Z time="2023-05-23T14:10:51Z" level=error msg="markers: did not save new marker for 2019/12/20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:51.607376332Z time="2023-05-23T14:10:51Z" level=info msg="index: found 2 faces in 20191211_140909_EF5BAC99.jpg [124.007601ms]"
> photoprism_1  | 2023-05-23T14:10:53.451181853Z time="2023-05-23T14:10:53Z" level=debug msg="photo: changing location of 2019/12/20191211_140909_EF5BAC99 from zz to s2:3bae0b219dfc"
> photoprism_1  | 2023-05-23T14:10:53.778594631Z time="2023-05-23T14:10:53Z" level=debug msg="photo: 2019/12/20191211_140909_EF5BAC99 title based on label 'The Empyrean'"
> photoprism_1  | 2023-05-23T14:10:53.778671318Z time="2023-05-23T14:10:53Z" level=debug msg="photo: 2019/12/20191211_140909_EF5BAC99 has new title 'The Empyrean / Jagadenahalli / 2019' [2.027553ms]"
> photoprism_1  | 2023-05-23T14:10:56.264843830Z time="2023-05-23T14:10:56Z" level=debug msg="search: updated index by photo id [360.884629ms]"
> photoprism_1  | 2023-05-23T14:10:56.270272165Z time="2023-05-23T14:10:56Z" level=debug msg="index: updated yaml file 20191211_140909_EF5BAC99.yml"
> photoprism_1  | 2023-05-23T14:10:56.270360248Z time="2023-05-23T14:10:56Z" level=info msg="import: added main jpg file 2019/12/20191211_140909_EF5BAC99.jpg"
> photoprism_1  | 2023-05-23T14:10:56.323644281Z time="2023-05-23T14:10:56Z" level=debug msg="exiftool: extracting metadata from Digikam_TO_BE_FILTERED/20191211_193912.jpg"
> photoprism_1  | 2023-05-23T14:10:56.928100969Z time="2023-05-23T14:10:56Z" level=debug msg="import: created 75076f4af309e03970a8dc93a645b6d59bf10139_exiftool.json"
> photoprism_1  | 2023-05-23T14:10:56.957007358Z time="2023-05-23T14:10:56Z" level=debug msg="metadata: successfully decoded Samsung trailer data for 20191211_193912.jpg"
> photoprism_1  | 2023-05-23T14:10:56.960946242Z time="2023-05-23T14:10:56Z" level=info msg="media: 20191211_193912.jpg was taken at 2019-12-11 14:09:12 +0000 UTC (meta)"
> photoprism_1  | 2023-05-23T14:10:56.964104854Z time="2023-05-23T14:10:56Z" level=info msg="import: moving main jpg file Digikam_TO_BE_FILTERED/20191211_193912.jpg to 2019/12/20191211_140912_036DA417.jpg"
> ```
> 
> I'm quite frequently getting this error: `did not save new marker for`

Please ignore this comment. I triggered 2 tests together, that of `import` and `This PR`. I'll retest with a fresh head tomorrow.

kvalev committed:
People: Import exif face regions for new files

The exif face regions import only worked for already indexed files, due to
erroneous assumption. For new files the file uid is determined at the very end
of the import process, when the file is saved. This means that no new markers
could be created from the exif face regions - instead a cryptic message was
logged. Deferring the marker creation to the end of the import process fixes the
issue.

kvalev committed:
People: Hardcode the score exif face regions

The hardcoded value is purposefully low to prevent the region from being used
for face clustering and auto-tagging.

kvalev committed:
People: Set marker name based on face region data

When reindexing an existing library, there might already be non-named face
detected markers. If this is the case and there is an overlapping marker in
exif, then the non-named face marker will be named based on the exif data.

kvalev committed:
People: Distinguish between exif and detected face markers

kvalev commented:
> I'm quite frequently getting this error: `did not save new marker for` 

@rickysarraf I was able to reproduce the error - it occurs when doing an index from scratch. The latest commits should fix the issue. I tried a bunch of different things (documented in the PR description) and couldn't encounter any issues - both fresh index and reindex as well as reindex after version update work fine. However if you can run a couple of tests yourself that would be great 👍

kvalev committed:
People: Remove superfluous embeddings check

rickysarraf commented:
> > I'm quite frequently getting this error: `did not save new marker for`
> 
> @rickysarraf I was able to reproduce the error - it occurs when doing an index from scratch. The latest commits should fix the issue. I tried a bunch of different things (documented in the PR description) and couldn't encounter any issues - both fresh index and reindex as well as reindex after version update work fine. However if you can run a couple of tests yourself that would be great +1

I think I can now safely conclude my tests.

In my opinion, this PR has now shaped into a fine feature. After a full scan, I'm yet to see any error. The only informative (error) message I see occasionally is: 

```
photoprism_1  | 2023-05-23T05:11:23.293727931Z time="2023-05-23T05:11:23Z" level=error msg="index: unexpected embeddings count 0 for face region crop area 1f42b1076042 and file 2023/05/20230520_
105505_C19CB773.jpg"
```

which I understand happens with low resolution photographs with much more lower resolution face embeddings.

Other than that, everything else I hoped for, is all working perfect. After the full scan, my PP instance is populated with 379 People, all done by this PR. 🙏🏽

kvalev commented:
Perfect, thanks a lot @rickysarraf!!

Locally I already have a follow-up for the `unexpected embeddings count` problem, however I am not sure how helpful it will be. The fallback boils down to running the face detection and check whether there is a sufficiently overlapping region with the tagged face. However it is debatable how many out of those zero-embeddings-faces would be found this way.

For now I will merge this PR as-is and if you are up for it, I can create a follow-up PR if you want to test it out.

Out of curiosity - do you know how many zero-embeddings-faces are there in your dataset, percentage wise? If you still have the logs from the indexing operation, you can simply count the number of `processing face region` and `unexpected embeddings count` logs. And have you checked some of them? Are they all very small faces like the one sample file you sent me previously?

rickysarraf commented:
> Perfect, thanks a lot @rickysarraf!!
> 
> Locally I already have a follow-up for the `unexpected embeddings count` problem, however I am not sure how helpful it will be. The fallback boils down to running the face detection and check whether there is a sufficiently overlapping region with the tagged face. However it is debatable how many out of those zero-embeddings-faces would be found this way.
> 
> For now I will merge this PR as-is and if you are up for it, I can create a follow-up PR if you want to test it out.
> 

Thank you for merging this one. Please feel free to create new PR and I'll then follow-up with tests.

> Out of curiosity - do you know how many zero-embeddings-faces are there in your dataset, percentage wise? If you still have the logs from the indexing operation, you can simply count the number of `processing face region` and `unexpected embeddings count` logs. And have you checked some of them? Are they all very small faces like the one sample file you sent me previously?

I don't have that information no more. I didn't save the entire log set.

But I have a scheduled full scan next week, which is when I will ensure to save the entire logs and provide you with the numbers.

rickysarraf commented:
> But I have a scheduled full scan next week, which is when I will ensure to save the entire logs and provide you with the numbers.

@kavlev

I have now finished all my testing.

I have, in total, `863` reports mentioning about embeddings.


```
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:41:49Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-3752-2318-505 to file frvmlnhoo72
9o6dx)"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:43:49Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-212-1596-280 to file frvmlthngqyyq
ise)"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:44:01Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-1008-1814-239 to file frvmluw2qkpl
1goi)"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:45:36Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face kiran-santosh-sarraf-2020-1560-80 to file frvmm232kjm804ti)"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:45:37Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face priya-raj-sarraf-2020-1289-60 to file frvmm232kjm804ti)"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:49:11Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1820-1489-303 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:49:21Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1254-1448-404 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T20:52:36Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1568-1746-694 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:00:24Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-2508-324-526 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:02:03Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face priya-raj-sarraf-2156-1753-694 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:02:58Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-2572-3020-1406 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:03:14Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1269-2017-873 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:03:20Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1591-1585-863 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:05:27Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face priya-raj-sarraf-1433-1320-1320 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:07:41Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-3999-2348-597 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:07:52Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face priya-raj-sarraf-656-2818-1250 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:08:26Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1275-2429-705 to file '')"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T21:11:18Z" level=error msg="index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-1620-1380-315 to file '')"
```

I wasn't too concerned about those nil file references because they may be the references to older files that were moved/renamed.

I like the fact that, now, PHotoPrism doesn't chew all cpu forever, when doing periodic face scans in the background. I think part of the reason is that it gives higher precedence to meta information.

```
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:36Z" level=info msg="faces: resolved 35 conflicts"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:36Z" level=warning msg="faces: marker mrvkzwi3it4qcmjk face CTR5IAGKT63LPJUAVULDUKGXCR5J2DTQ has ambiguous subjects jrvkzwd1vsadmyvb <> jrvkzup1qs9rlt42, subject source meta"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:37Z" level=info msg="faces: resolved 9 conflicts"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:37Z" level=warning msg="faces: marker mrvkzwp4zibrbw5b face 5CEDCOB6LL4WX63CBY7A2YCCGEMNCVY3 has ambiguous subjects jrvkzup1qs9rlt42 <> jrvkzuhd284yl1ql, subject source meta"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:41Z" level=info msg="faces: resolved 33 conflicts"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:41Z" level=warning msg="faces: marker mrvkzxui9yu0bt2o face IQGQC775UCT3LF4AFVTBW7UKFY52ZOY6 has ambiguous subjects jrvkzup1qs9rlt42 <> jrvkzv03ld9vdq0c, subject source meta"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:42Z" level=info msg="faces: resolved 10 conflicts"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:42Z" level=warning msg="faces: marker mrvkzy4bdrexdexr face QMHVFX6LJSCYUSSHB7I6PN5FC7BTVHBD has ambiguous subjects jrvkzxx3gfuzuh09 <> jrvkzx13n8v0xg2m, subject source meta"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:52Z" level=info msg="faces: resolved 2 conflicts"
ESC[33mphotoprism_1  |ESC[0m time="2023-06-02T12:58:52Z" level=warning msg="faces: marker mrvl063bphm51qr1 face NXTFSBX5KVMGX5CPFOUK2VN7OUX5TI2V has ambiguous subjects jrvkzuq18bljxcnm <> jrvkzvt153f50uea, subject source meta"
```

kvalev commented:
Thanks!

Those zero embeddings for the `''` files are quite interesting, as the face regions are quite large, so detection should have worked fine.

I have cleaned up the fallback implementation a bit, but I had no data to test it, so it might be buggy. If you want you can give it ago - https://github.com/kvalev/photoprism/pull/135. I can test and debug it locally, but once again I will need few sample photos - the first 3 files from your log (especially the first one) have large enough faces for the fallback to work and are prime candidates for testing.

```
index: face does not have exactly one embedding, has 0 (adding face anshika-raj-sarraf-3752-2318-505 to file frvmlnhoo72
9o6dx)
index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-212-1596-280 to file frvmlthngqyyq
ise)
index: face does not have exactly one embedding, has 0 (adding face ritesh-raj-sarraf-1008-1814-239 to file frvmluw2qkpl
1goi)
```
